sdk/attestation/azure-security-attestation/azure/security/attestation/_administration_client.py

```diff
@@ -0,0 +1,212 @@
+        """ Returns the base URL configured for this instance of the AttestationClient.
+
+        :returns str: The base URL for the client instance.
+        """
+        return self._base_url
+
+    @distributed_trace
+    def get_policy(self, attestation_type, **kwargs): 
+        #type(AttestationType) -> AttestationResult[str]:
+        """ Retrieves the attestation policy for a specified attestation type.
+
+        :param azure.security.attestation.AttestationType attestation_type: :class:`azure.security.attestation.AttestationType` for 
+            which to retrieve the policy.
+        :return AttestationResponse[str]: Attestation service response encapsulating a string attestation policy.
+        """
+        
+        policyResult = self._client.policy.get(attestation_type, **kwargs)
+        token = AttestationToken[PolicyResult](token=policyResult.token, body_type=PolicyResult)
+        token_body = token.get_body()
+        stored_policy = AttestationToken[StoredAttestationPolicy](token=token_body.policy, body_type=StoredAttestationPolicy)
+
+        actual_policy = stored_policy.get_body().attestation_policy #type: bytes
+
+        if self._config.token_validation_options.validate_token:
+            token.validate_token(self._config.token_validation_options, self._get_signers(**kwargs))
+
+        return AttestationResponse[str](token, actual_policy.decode('utf-8'))
+
+    @distributed_trace
+    def set_policy(self, attestation_type, attestation_policy, signing_key=None, **kwargs): 
+        #type:(AttestationType, str, AttestationSigningKey, Any) -> AttestationResponse[PolicyResult]
+        """ Sets the attestation policy for the specified attestation type.
+
+        :param azure.security.attestation.AttestationType attestation_type: :class:`azure.security.attestation.AttestationType` for 
+            which to set the policy.
+        :param str attestation_policy: Attestation policy to be set.
+        :param AttestationSigningKey signing_key: Optional signing key to be
+            used to sign the policy before sending it to the service.
+        :return AttestationResponse[PolicyResult]: Attestation service response encapsulating a :class:`PolicyResult`.
+        """
+        policy_token = AttestationToken[StoredAttestationPolicy](
+            body=StoredAttestationPolicy(attestation_policy = attestation_policy.encode('ascii')),
+            signer=signing_key,
+            body_type=StoredAttestationPolicy)
+        policyResult = self._client.policy.set(attestation_type=attestation_type, new_attestation_policy=policy_token.serialize(), **kwargs)
+        token = AttestationToken[PolicyResult](token=policyResult.token,
+            body_type=PolicyResult)
+        if self._config.token_validation_options.validate_token:
+            if not token.validate_token(self._config.token_validation_options, self._get_signers(**kwargs)):
+                raise Exception("Token Validation of PolicySet API failed.")
```

Comment:
This one is tough. There are three possible causes of this particular error:

1.  The customers validation callback returned False.
2.  The token returned from the caller is expired (or otherwise doesn't meet the criteria (too early, too late)).
3. Something went horribly wrong with the service and the token's signature is invalid.

For  .Net we defined a custom "token validation failed exception", do you recommend something similar for Python?

Reply:
I think that is a good solution, and passing in a message about what the issue was would be awesome for the user.